File: go/4.go

package main

import (
	"fmt"
	"strconv"
)

func palindrome(num int) int {
	s := strconv.Itoa(num)
	l := len(s)
	for i := 0; i < l / 2; i += 1 {
		if s[i] != s[l - i - 1] {
			return 0
		}
	}
	return num
}

func main() {
	start, end := 999, 99
	lp, num, p := 0, 0, 0
	for i := start; i > end; i = i - 1 {
		for j := start; j > end; j = j - 1 {
			num = i * j
			p = palindrome(num)
			if p > 0 && p > lp {
				lp = p
			}
		}
	}
	fmt.Println("The largest palindrome made from the product of two 3-digit numbers is", lp)
}

File: py/4.py
#!/usr/bin/env python

import sys

def palindrome(s_num):
    length = len(s_num)
    for x in range(length / 2):
        if s_num[x] is not s_num[length - x - 1]:
            return False
    return s_num


if __name__ == '__main__':
    start = 999
    end = 99
    lp = 0
    for i in range(start, end, -1):
        for j in range(start, end, -1):
            num = i * j
            p = int(palindrome(str(num)))
            if p and p > lp:
                lp = p

    print >> sys.stdout, "The largest palindrome made from the product of two 3-digit numbers is: %s" % lp
